### PR TITLE
String function implementation

### PR DESCRIPTION
**Pull request by bfinzer**
[#187526964] Implement V2 string functions.

* Added two new "utils" files, one for data and one for dates. These are utilities needed for implementing string functions
* Implemented V2 string functions except for `combine` and `wordListMatches`
* Provided string-functions.test.ts

### CODE REVIEW

File: v3/src/models/formula/functions/string-functions.ts [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    numOfRequiredArguments: 2,
+    evaluate: (...args: FValue[]) => {
+      const text = valueToString(args[0]), suffix = valueToString(args[1])
+      if (!suffix) return true
+      if (!text) return false
+      return text.endsWith(suffix)
+    }
+  },
+  // findString(stringToLookIn, stringToFind, start) Takes three arguments, returning the position of
+  // stringToFind in stringToFind starting from start. The first character of stringToLookIn is numbered 1.
```
Comment (kswenson) [inline v3/src/models/formula/functions/string-functions.ts:49]:
```suggestion
  // stringToFind in stringToLookIn starting from start. The first character of stringToLookIn is numbered 1.
```

File: v3/src/models/formula/functions/string-functions.ts [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    }
+  },
+  // patternMatches(stringToLookIn, pattern) Returns the number of times in string_to_look_in the string or
+  // regular expression represented by pattern occurs. The search is case-insensitive.
+  patternMatches: {
+    numOfRequiredArguments: 2,
+    evaluate: (...args: FValue[]) => {
+      const stringToLookIn = valueToString(args[0]),
+        pattern = valueToString(args[1])
+      return (stringToLookIn.match(new RegExp(pattern, "gmi")) || []).length
```
Comment (kswenson) [inline v3/src/models/formula/functions/string-functions.ts:86]: Strings, particularly user-entered strings, should always be escaped before using them as regular expressions.
```suggestion
      return (stringToLookIn.match(new RegExp(escapeStringRegexp(pattern), "gmi")) || []).length
```

File: v3/src/models/formula/functions/string-functions.ts [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+  },
+  // replaceString(aString, stringToFind, substituteString) Takes three string arguments and substitutes the third
+  // for all occurrences of the second in the first.
+  replaceString: {
+    numOfRequiredArguments: 3,
+    evaluate: (...args: FValue[]) => {
+      const aString = valueToString(args[0]),
+        stringToFind = valueToString(args[1]),
+        substituteString = valueToString(args[2])
+      return aString.replace(new RegExp(stringToFind, "g"), substituteString)
```
Comment (kswenson) [inline v3/src/models/formula/functions/string-functions.ts:121]: Strings, particularly user-entered strings, should always be escaped before using them as regular expressions.
```suggestion
      return aString.replace(new RegExp(escapeStringRegexp(stringToFind), "g"), substituteString)
```

File: v3/src/utilities/data-utils.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import { isDate } from "mathjs"
+
+export const valueToString = function (iValue: any): string {
+  let valType:string = typeof iValue,
```
Comment (kswenson) [inline v3/src/utilities/data-utils.ts:4]:
```suggestion
  let valType: string = typeof iValue,
```

File: v3/src/utilities/data-utils.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    value = iValue
+  if (isDate(iValue)) {
+    // treat dates as strings
+    // todo: value = formatDate(iValue)
+    value = String(iValue)
+    valType = "string"
+  } else if (iValue instanceof Error) {
+    value = `${iValue.name  } ${  iValue.message}`
```
Comment (kswenson) [inline v3/src/utilities/data-utils.ts:12]:
```suggestion
    value = `${iValue.name} ${iValue.message}`
```

**Review by kswenson — CHANGES_REQUESTED** [review]
👍 Generally looks good. One suggestion to use `escapeStringRegexp` when converting strings to regular expressions. Beyond that, I think we'll need some additional argument validation for many of these functions. What happens when arguments that should be numbers are `NaN`s or other non-numeric values? What happens when real numbers or negative numbers are passed when indices are expected?

This would be a good opportunity to try some test-driven development (TDD), .e.g.:
- write a test for what should happen when 1.5 is passed as an index and then update the code (if necessary) to pass the test.
- write a test to search for a `.` in a string that doesn't have one. (I would expect the test to fail because the search should succeed since `.` is a wildcard in regular expressions.) Then add the call to `escapeStringRegexp` I recommended above which should allow the test to pass. 

The tests for each function could be expanded to check these edge cases and validate the arguments appropriately.

**Review by kswenson — APPROVED** [review]
👍 LGTM
